### ISSUE TITLE
static variable causes typescript lowering

### ISSUE THREAD

**Issue by hyrious** [evanw/esbuild#1480]
I've noticed lowering always occurs in `--target=esnext` when there is static variable in class and the loader is `ts` or `tsx` (`js` and `jsx` are not affected). For example, this code:

```js
class A {
    static C = 1;
    #a;
    f() { this.#a ??= A.C }
}
```

will always be lowered to something include `__privateGet`. while this code is not lowered when the loader is `js`.

Is this an intentional behavior?

Related issue: https://github.com/evanw/esbuild/issues/1418#issuecomment-886652553

[Live example](https://hyrious.me/esbuild-repl/?version=0.12.18&shareable=JTdCJTIyY29kZSUyMiUzQSUyMmNsYXNzJTIwQSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBzdGF0aWMlMjBDJTIwJTNEJTIwMSUzQiU1Q24lMjAlMjAlMjAlMjAlMjNhJTNCJTVDbiUyMCUyMCUyMCUyMGYoKSUyMCU3QiUyMHRoaXMuJTIzYSUyMCUzRiUzRiUzRCUyMEEuQyUyMCU3RCU1Q24lN0QlMjIlMkMlMjJjb25maWclMjIlM0ElN0IlMjJsb2FkZXIlMjIlM0ElMjJ0cyUyMiUyQyUyMnRhcmdldCUyMiUzQSUyMmVzbmV4dCUyMiU3RCU3RA%3D%3D)

**Comment by evanw** [comment]
I believe in this case the issue is that the official TypeScript compiler defaults to [`useDefineForClassFields: false`](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) so esbuild does too. When you set it to true such as with this:

```
--loader=ts --target=esnext --tsconfig-raw={"compilerOptions":{"useDefineForClassFields":true}}
```

then that code is not lowered. Specifically when `useDefineForClassFields` is `false` the public class field initializer `static C = 1;` must be transformed to `A.C = 1;` to change from "define" to "assign" semantics. When that happens, the private class field initializer `#a` is lowered as well in case the initializer for `C` references it. This is necessary because the initializer is moved outside of the class body and syntactically the symbol `#a` is only valid inside of the class body. For example, if the variable was initialized to `static C = new A().#a;` instead, it would be a syntax error to convert that to `A.C = new A().#a;` outside of the class body. This check is somewhat conservative but it's done for the correctness of the resulting code. If you want this exact pass-through behavior, you'll have to configure TypeScript such that class fields behave like JavaScript since it's not the default behavior.

**Comment by hyrious** [comment]
thank you for so detailed answer.
maybe we can make `useDefineForClassFields` default to true when using `transform` with target=esnext?

**Comment by hyrious** [comment]
Update: I found typescript has [a simple transform api](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#a-simple-transform-function), while its default behavior is targetting commonjs and es5.

```js
const ts = require('typescript')

let start = performance.now()
let result = ts.transpileModule(`
  class A { a = 1 }
`, {
  compilerOptions: {
    target: 99 // esnext, if not set this, it will output es5
  }
})
let elapsed = performance.now() - start

console.log(result.outputText)
console.log(`[Finished in ${elapsed.toFixed(2)}ms]`)
```

Output:
```js
class A {
    a = 1;
}

[Finished in 36.37ms]
```

**Comment by hyrious** [comment]
Since it was resolved in 0.14.0 (you can use `--target=esnext` to enable `useDefineForClassFields` in transform mode), closing it.
